File: 424-longest-repeating-character-replacement/longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        int[] arr = new int[26];
        int i = 0;
        int max = 0;
        int j = 0;
        int result = 0;
        while(i < s.length()){
            arr[s.charAt(i) - 'A']++;
            max = Math.max(max, arr[s.charAt(i) - 'A']);
            int window = i - j + 1;
            int replacements = window - max;
            while(replacements > k){
                arr[s.charAt(j)-'A']--;
                j++;
                window = i - j + 1;
                replacements = window - max;
            }
            result = Math.max(result, i-j+1);
            i++;
        }
        return result;
    }
}

File: 424-longest-repeating-character-replacement/longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        hashmap = {}
        i = 0
        j = 0
        currmax = 0
        result = 0
        for i in range(0, len(s)):
            if s[i] in hashmap:
                hashmap[s[i]] += 1
            else:
                hashmap[s[i]] = 1
            currmax = max(currmax, hashmap[s[i]])
            if ((i - j + 1) - currmax) > k:
                if hashmap[s[j]] == 1:
                    hashmap.pop(s[j])
                else:
                    hashmap[s[j]] -= 1
                j = j + 1
            result = max(result, i - j + 1)
            i = i + 1
        return result






